python/ModelTools.py

```diff
@@ -97,6 +97,23 @@ def doModel(self):
                 print "Wrote GraphVizTree of model_s to ",self.options.out+".dot"
 
     def doRateParams(self):
+
+    	# First support external functions/parameters 
+	for rp in self.DC.rateParams.keys():
+	 for rk in range(len(self.DC.rateParams[rp])):
+	  type = self.DC.rateParams[rp][rk][0][-1]
+	  if type!=2: continue
+	  argu,argv = self.DC.rateParams[rp][rk][0][0],self.DC.rateParams[rp][rk][0][1]
+	  if self.out.arg(argu): continue
+	  fin,wsn = argv.split(":")
+	  try:
+	    fitmp = ROOT.TFile.Open(fin)
+	    wstmp = fitmp.Get(wsn)
```

Comment:
Yes, thats a good idea. We could keep a map of the files/wspaces open and
re-use them, with a close at the end.

On Mon, Mar 21, 2016 at 2:48 PM, Andrew Gilbert <notifications@github.com>
wrote:

> In python/ModelTools.py
> <https://github.com/cms-analysis/HiggsAnalysis-CombinedLimit/pull/290#discussion_r56823733>
> :
>
> > @@ -97,6 +97,23 @@ def doModel(self):
> >                  print "Wrote GraphVizTree of model_s to ",self.options.out+".dot"
> >
> >      def doRateParams(self):
> > +
> > +    	# First support external functions/parameters
> > +	for rp in self.DC.rateParams.keys():
> > +	 for rk in range(len(self.DC.rateParams[rp])):
> > +	  type = self.DC.rateParams[rp][rk][0][-1]
> > +	  if type!=2: continue
> > +	  argu,argv = self.DC.rateParams[rp][rk][0][0],self.DC.rateParams[rp][rk][0][1]
> > +	  if self.out.arg(argu): continue
> > +	  fin,wsn = argv.split(":")
> > +	  try:
> > +	    fitmp = ROOT.TFile.Open(fin)
> > +	    wstmp = fitmp.Get(wsn)
>
> I guess this could be slow if these input workspaces are large and/or
> there are many rateParam lines all needing something from the same
> workspace. We could probably keep the workspaces open for the duration of
> this function in this case, but could be implemented later
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-analysis/HiggsAnalysis-CombinedLimit/pull/290/files/2d64e9d84ff66d4f1225dc107c8a4f1380da12a3#r56823733>
>